### ISSUE TITLE
Will Prisma 2 be adding support for multiple databases?

### ISSUE THREAD

**Issue by astr0sl0th** [prisma/prisma2#1274]
As previously stated in the v1 docs multiple database support isn't currently available but there are plans on including it in the future is this a feature that will be included in Prisma 2?

**Comment by nikolasburk** [comment]
Hey @astr0sl0th, thanks a lot for opening this issue! Adding support for multiple databases is certainly on our radar and is something that we'll tackle eventually. However, we can't commit to any timeline for this yet and it's unlikely (though not impossible) that we'll launch anything this year.

It would be great if you could share a bit more info about your concrete use case (e.g. which DBs you want to combine and why, what your requirements are, ...). This helps us a lot in prioritizing and refining the features we are going to build.

Note that we'll keep this issue open as a tracking issue fo supporting multiple database (I also took the liberty of updating the title to make this a bit more clear).

**Comment by donedgardo** [comment]
I think it would be a great feature, not to merge database schemas but possibly having two different instances of prisma running on the same service, and it would work like different datasources each with its own typings and migrations on its own.

I understand this could probably be already done, although I haven't tried it. Maybe outputting the source of the prisma client to different generated folders. 

This goes to say that having a federated apollo server would probably work when each service has its own primsa instance for the database it needs, but having two or multiple instances for when you have different databases and you are not working on a federated architecture.

**Comment by LeoFalco** [comment]
I'm interested in this feature too

**Comment by davidboom95** [comment]
+1

**Comment by willwillis** [comment]
I would like this feature as well. My use case is to migrate data in a legacy mysql database to a new "prisma schema" postgres database.  I'd like to generate clients for each and read data from one to insert into the other. Is this not possible due to the client being generated in a common location? (node_modules/@prisma/client)

**Comment by wuntusk** [comment]
Same problem.

**Comment by astr0sl0th** [comment]
Excuse the delay in reply to my own thread but I have been on a sabbatical.

My main use case for this feature was an exit strategy on per brand basis. Although the databases were identical in structure being Postgres each brand needed its own separate database and currently combining multiple databases was not an option which is why I asked if this was going to be a feature in Prisma 2.

**Comment by nclsndr** [comment]
+1

**Comment by mingchuno** [comment]
+1

**Comment by PerfectedApp** [comment]
+1

**Comment by nafkhanzam** [comment]
+1

**Comment by gcherem** [comment]
+1

**Comment by dfee** [comment]
Just wanted to come here and say this actually works if you:
- use a monorepo, and separate out each databases into its own package
- specify a custom outpath as [described here](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/generating-prisma-client#specifying-a-custom-output-path)
- add  `"main": "dist/index.js",` and `"types": "dist/index.d.js",` to your `package.json`.
- use a `build` script like: `"build": "prisma generate"`

Then, when you them as a dependencies in your other package, you can use them like:
```
import { PrismaClient as PrismaOneClient } from "@my-org/db-one";
import { PrismaClient as PrismaTwoClient } from "@my-org/db-two";

process.env.DB_ONE_URL = "postgresql://dfee:dfee@localhost:5432/db-one?schema=public";
process.env.DB_TWO_URL = "postgresql://dfee:dfee@localhost:5432/db-one?schema=public";

const client1 = new PrismaOneClient();
const client2 = new PrismaTwoClient();

const main = async () => {
  const allUsers = await client1.user.findMany();
  console.log(allUsers);
  const allAnimals = await client2.animal.findMany();
  console.log(allAnimals);
};

main()
    .then((v) => console.log("done"))
    .catch((e) => {
      console.log(e);
    })
    .finally(async () => {
      await client.$disconnect();
      await client2.$disconnect();
    });
```

**Comment by blazestudios23** [comment]
@dfee I had the same thought as you and did what you did. Unfortunately I can't get Nexus to recognize my clients. I'm not sure what's going to there. It just says I never ran Prisma Generate. I will update when I figure it out.

**Comment by blazestudios23** [comment]
So I have a usecase where I need to connect to multiple schemas on a MySQL database. May company deals with massive amounts of DB tables and its easiest to organize these tables into schema for related content. Each DB can have up to 100 tables. 

The best solution would be able to hit multiple schemas and combine them into one Prisma Client.

**Comment by alnimra** [comment]
> @dfee I had the same thought as you and did what you did. Unfortunately I can't get Nexus to recognize my clients. I'm not sure what's going to there. It just says I never ran Prisma Generate. I will update when I figure it out.

Did you figure this out?

**Comment by dfee** [comment]
I think this should be a separate issue: "stitching multiple schemas into one client".

**Comment by blazestudios23** [comment]
@alnimra  So far there is no way to solve this problem. But GraphQL is the exact tech one would use to stitch DBs together, so this something that should be resolved.

**Comment by blazestudios23** [comment]
@dfee Yes, it may be a separate ticket, I feel that having multiple clients is probably not the best solution though, it just makes more sense to stitch all DBs into one client.

**Comment by m-leon** [comment]
@dfee 
Do you have an example repo with the monorepo setup?

**Comment by CatsMiaow** [comment]
https://github.com/prisma/prisma/issues/2443#issuecomment-630679118

**Comment by blazestudios23** [comment]
So I did figure out a solution using Nexus Prisma plugin where I stitched the schemas together that were generated by nexus.

**Comment by BetterLuke** [comment]
> Just wanted to come here and say this actually works if you:
> 
> * use a monorepo, and separate out each databases into its own package
> * specify a custom outpath as [described here](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/generating-prisma-client#specifying-a-custom-output-path)
> * add  `"main": "dist/index.js",` and `"types": "dist/index.d.js",` to your `package.json`.
> * use a `build` script like: `"build": "prisma generate"` ([alternatively, you could use a pre-install hook](https://docs.npmjs.com/misc/scripts))
> 
> Then, when you them as a dependencies in your other package, you can use them like:
> 
> ```ts
> import { PrismaClient as PrismaOneClient } from "@my-org/db-one";
> import { PrismaClient as PrismaTwoClient } from "@my-org/db-two";
> 
> process.env.DB_ONE_URL = "postgresql://dfee:dfee@localhost:5432/db-one?schema=public";
> process.env.DB_TWO_URL = "postgresql://dfee:dfee@localhost:5432/db-one?schema=public";
> 
> const client1 = new PrismaOneClient();
> const client2 = new PrismaTwoClient();
> 
> const main = async () => {
>   const allUsers = await client1.user.findMany();
>   console.log(allUsers);
>   const allAnimals = await client2.animal.findMany();
>   console.log(allAnimals);
> };
> 
> main()
>     .then((v) => console.log("done"))
>     .catch((e) => {
>       console.log(e);
>     })
>     .finally(async () => {
>       await client1.$disconnect();
>       await client2.$disconnect();
>     });
> ```

Hi dfee, Are you use nx to make monorepo?

**Comment by dfee** [comment]
No. I use `pnpm`. But, I'm not sure how much the package / monorepo manager actually matters.

**Comment by BetterLuke** [comment]
> No. I use `pnpm`. But, I'm not sure how much the package / monorepo manager actually matters.

I found a temporary solution, see here: 😉 https://github.com/prisma/prisma/issues/7470#issuecomment-873598065

**Comment by AleksandrukTad** [comment]
Hello got the same problem'
I'd like to create "shared-service" which would be common data access layer between express app (postgres) and electron app (sqlite). I would use "shared-service" as npm package and ideally just pass the provider name and url to connect it. I need this for offline support (fallback to sqlite) for electron app

**Comment by BetterLuke** [comment]
> 您好，遇到了同样的问题，
> 我想创建“共享服务”，这将是 express 应用程序（postgres）和电子应用程序（sqlite）之间的通用数据访问层。我会使用“共享服务”作为 npm 包，理想情况下只传递提供者名称和 url 来连接它。我需要这个用于电子应用程序的离线支持（回退到 sqlite）

I saw an article yesterday, not sure if it can help you: https://darioielardi.dev/schema-based-multi-tenancy-with-nestjs-and-prisma?utm_source=pocket_mylist

**Comment by AmreeshTyagi** [comment]
> > 您好，遇到了同样的问题，
> > 我想创建“共享服务”，这将是 express 应用程序（postgres）和电子应用程序（sqlite）之间的通用数据访问层。我会使用“共享服务”作为 npm 包，理想情况下只传递提供者名称和 url 来连接它。我需要这个用于电子应用程序的离线支持（回退到 sqlite）
> 
> I saw an article yesterday, not sure if it can help you: https://darioielardi.dev/schema-based-multi-tenancy-with-nestjs-and-prisma?utm_source=pocket_mylist

only suitable for small number of tenant. Not a cost effective solution, if you want to do multi-tenancy only at database level not at application level.

**Comment by janpio** [comment]
Potential duplicate of https://github.com/prisma/prisma/issues/2443

**Comment by lightningspirit** [comment]
Based on my experience with Prisma and looking to past use-cases where multiple databases where used, I post here my suggestion on the possible schema to handle different datasources. The following example uses a MySQL for transactional / schema enforced models (e.g. Product, User, Account, Payment, etc) and a MongoDB instance for Analytical models (e.g. View, Event, etc).

### `@@datasource` attribute
A new `@@datasource` attribute is available to be included in every model (required if schema has more than one datasource). Any number of datasources is allowed in the schema.

### Relationships
Any model can participate in any kind of relationship with other models (i.e. one-to-one, one-to-many, many-to-many). If explicit joint tables are provided in the schema, they also must have `@@datasource`, which means, the user must explicitly tell Prisma in which datasource the joint table is created.

Despite of `relationMode` being explicitly provided in a datasources, relation keys between different datasources must be enforced by the Prisma engine, i.e. `relationMode = "prisma"` (e.g. in `Event` collection, the existence of `productId` in `Product` table must be ensured by Prisma engine).

### Example:
```prisma
generator client {
  provider     = "prisma-client-js"
}

datasource transactional {
  provider     = "mysql"
  url          = env("TRANSATIONAL_DATABASE_URL")
}

datasource analytics {
  provider     = "mongodb"
  url          = env("ANALYTICS_DATABASE_URL")
}

model Product {
  id            String            @id @default(cuid())
  events        Event[]
  ...
  @@datasource("transactional")
}

enum EventType {
  View
  ...
}

model Event {
  id            String            @id @default(auto()) @map("_id") @db.ObjectId
  type          EventType
  product       Product           @relation(field: [productId], references: [id])
  productId     String
  ...
  @@index([type])
  @@index([productId])
  @@datasource("analytics")
}
```
I hope this comment can be useful to put this VERY useful feature forward in the next versions. I'm available to point out more use-cases / info if needed.

**Comment by souzacavalheiro** [comment]
Hello everybody! I didn't see in recent comments any use case related to 'db for user group'.

exemplifying:

  userA => group1 => mysqldb1
  userb => group2 => mysqldb2

Our users are company employees and each one accesses their own database... same tables, same schema, same provider but different databases.
Any tips?

ps: translated by google

**Comment by amilcarrey** [comment]
Is there any news on this? @matthewmueller
